torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+    }
+  }
+  void collect(const c10::TensorOptions& t) {
+    // TODO(jansel): are there any pointers in this type we shouldn't memcmp
+    // I think this one is wrong.... should fix it
+    specialize_on_bytes(t);
+  }
+  void collect(const at::TensorGeometry& t) {
+    collect(t.sym_sizes());
+    collect(t.sym_strides());
+    collect(t.sym_storage_offset());
+  }
+  void collect(torch::autograd::generated::TypeAndSize& t);
+  void collect(const OutputRef& t) {
+    // +1 is for undefined encoded as -1
+    collect_size(t.node_id + 1);
+    collect_size(t.index + 1);
+  }
+  void collect(const NodeCall& t) {
+    collect(t.input_refs);
+    collect(t.graph_output);
+  }
+  void collect(int8_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(int16_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(int32_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(int64_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint8_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint16_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint32_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint64_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(bool t) {
+    specialize_on_bytes(t);
+  }
+  void collect(float t) {
+    specialize_on_bytes(t);
+  }
+  void collect(double t) {
+    specialize_on_bytes(t);
+  }
+
+  void set_grad_target(const at::Tensor& tensor) {
+    collect(_compiler.accumulate_grad);
+    if (_compiler.accumulate_grad)
+      _compiler.add_set_grad_target(tensor);
+  }
+
+  void collect_hooks_from(Node* fn) {
+    for (auto& i : fn->tensor_pre_hooks()) {
+      i->compiled_args(*this);
+    }
+    for (auto& i : fn->retains_grad_hooks()) {
+      i.second->compiled_args(*this);
+    }
+    for (auto& i : fn->pre_hooks()) {
+      i->compiled_args(*this);
+    }
+    for (auto& i : fn->post_hooks()) {
+      i->compiled_args(*this);
+    }
+    specialize_on_hook_counts();
+  }
+
+  void specialize_on_hook_counts() {
+    collect_size(_node_call.tensor_pre_hooks.size());
+    for (const auto& h : _node_call.tensor_pre_hooks) {
+      collect_size(h.second);
+    }
+    collect_size(_node_call.pre_hooks.size());
+    collect_size(_node_call.post_hooks.size());
+  }
+
+  CacheKey key() const {
+    Node* node = _node_call.node.get();
+    return CacheKey(
+        typeid(*node), _specialization_key, _specialization_key_size);
+  }
+
+  void add_tensor_pre_hook(c10::SafePyObject&& obj, int index) {
+    auto fn_id = _compiler.emplace_hook(std::move(obj));
+    collect_size(fn_id);
+    _node_call.tensor_pre_hooks.emplace_back(std::make_pair(fn_id, index));
+  }
+
+  void add_pre_hook(c10::SafePyObject&& obj) {
+    auto fn_id = _compiler.emplace_hook(std::move(obj));
+    collect_size(fn_id);
+    _node_call.pre_hooks.emplace_back(fn_id);
+  }
+
+  void add_post_hook(c10::SafePyObject&& obj) {
+    auto fn_id = _compiler.emplace_hook(std::move(obj));
+    collect_size(fn_id);
```

Comment:
It's not... entirely clear to me how you are checking that the hooks are actually the same from run to run? Here all you collect is the `fn_id`, but eventually we also have to check the PyObjects are the same right? Maybe that is checked special on pre_hooks/post_hooks, haven't gotten there yet.

Reply:
We aren't.  We lift hooks into args to the FX graph, and only specialize on the number of hooks (not their values).

We are relying on dynamo guards to check the hooks match.

Eventually we will need to support hooks that are defined as `lamba:` functions.  Which results in a different function object every call.  I think this can be handled at the dynamo level, which is why I made hooks args to the FX graph.